src/interactions/brush.js

```diff
@@ -0,0 +1,138 @@
+      const svg = context.ownerSVGElement;
+      const {data} = context.getMarkState(this);
+
+      // Isolate state per-plot.
+      let state = states.get(svg);
+      if (state && !index.fi) throw new Error("The brush interaction currently supports only one brush per plot.");
+      if (!state) {
+        // Derive intersection bounds.
+        const {x, y} = scales;
+        const bx = x?.bandwidth?.() ?? 0;
+        const by = y?.bandwidth?.() ?? 0;
+        const {x: X, x1: X1, x2: X2, y: Y, y1: Y1, y2: Y2} = values;
+        const Xl = X1 && X2 ? X1.map((d, i) => Math.min(d, X2[i])) : X;
+        const Xm = X1 && X2 ? X1.map((d, i) => Math.max(d, X2[i]) + bx) : bx ? X.map((d) => d + bx) : X;
+        const Yl = Y1 && Y2 ? Y1.map((d, i) => Math.min(d, Y2[i])) : Y;
+        const Ym = Y1 && Y2 ? Y1.map((d, i) => Math.max(d, Y2[i]) + by) : by ? Y.map((d) => d + by) : Y;
+
+        // This brush is shared by all the facets.
+        const {width, height, marginLeft, marginTop, marginRight, marginBottom} = dimensions;
+        const extent = [
+          [marginLeft, marginTop],
+          [width - marginRight, height - marginBottom]
+        ];
+        const brush = (mode === "xy" ? brusher : mode === "x" ? brusherX : brusherY)()
+          .extent(extent)
+          .on("brush start end", function (event, d) {
+            const {type, selection} = event;
+            let S = null;
+            const [X, Y] = !selection
+              ? []
+              : mode === "xy"
+              ? transpose(selection)
+              : mode === "x"
+              ? [selection]
+              : [, selection];
+            if (X || Y) S = d.index;
+            if (X) {
+              const [x0, x1] = X;
+              S = S.filter((i) => x0 <= Xm[i] && Xl[i] <= x1);
+            }
+            if (Y) {
+              const [y0, y1] = Y;
+              S = S.filter((i) => y0 <= Ym[i] && Yl[i] <= y1);
+            }
+            // Only one facet can be active at a time; clear the others.
+            if (type === "start") for (const target of targets) if (target !== this) target._cancelBrush();
+
+            d.display?.replaceWith((d.display = next.call(this, S ?? [], scales, values, dimensions, context)));
+
+            // Update the plot’s value if the selection has changed.
+            if (selectionMode === "data") {
+              if (!selectionEquals(S, state.selection)) {
+                state.selection = S;
+                context.dispatchValue(S === null ? data : take(data, S));
+              }
+            }
+            // "extent"
+            else {
+              if (selection === null) {
+                context.dispatchValue(null);
+              } else {
+                const value = {};
+                if (X) value.x = x.invert ? X.map(x.invert).sort(ascending) : X;
+                if (Y) value.y = y.invert ? Y.map(y.invert).sort(ascending) : Y;
```

Comment:
I want to think through the design here a bit more, and in particular, how we’d handle brushing over ordinal (point and band) scales. In the continuous case, I thought we might expose the selection using _x1_ and _x2_ properties, though saying _x_ = [_x1_, _x2_] also makes sense. In the ordinal case, I suppose we could use the same representation, but it would be more natural to use _x_ = [… the subset of values from the _x_ scale’s domain that are selected…].